text/0000-scoped-take-2.md

```diff
@@ -0,0 +1,262 @@
+were explicitly collected into a vector (to ensure that the joins did not happen
+too early).
+
+With the new version, by contrast, the scope is more clearly marked by an indent,
+all *all* joins within the scope automatically happen at the end of the block,
+making scoped threads feel more like a first-class control-flow construct.  In
+this RFC author's opinion, this actually winds up *clarifying* the semantics of
+scoped threads, and so may be a better API than the original `thread::scoped`.
+
+On a separate note, in practice one will usually not want to spawn full-fledged
+threads when doing data-parallel, fork-join style computations; instead you want
+to use a thread pool, work stealing, and higher-level combinators. One nice
+aspect of `Scope` is that crates providing such parallelism frameworks can
+easily hook into the API with their own means of spawning lightweight tasks;
+this should also facilitate a [decoupling of thread pools from the lifetimes
+bounding the tasks they are running](https://github.com/rust-lang/threadpool/issues/7).
+
+## Implementation
+
+Here's a sketch of the implementation, inspired in part by
+[@arielb1's ideas](https://gist.github.com/arielb1/5eb299a87546ce8829b3):
+
+```rust
+pub struct Scope<'a> {
+    dtors: RefCell<Option<DtorChain<'a>>>
+}
+
+struct DtorChain<'a> {
+    dtor: Box<FnBox() + 'a>,
+    next: Option<Box<DtorChain<'a>>>
+}
+
+pub fn scope<'a, F, R>(f: F) -> R where F: FnOnce(&Scope<'a>) -> R {
+    let mut scope = Scope { dtors: RefCell::new(None) };
+    let ret = f(&scope);
+    scope.drop_all();
+    ret
+}
+
+impl<'a> Scope<'a> {
+    // This method is carefully written in a transactional style, so
+    // that it can be called directly and, if any dtor panics, can be
+    // resumed in the unwinding this causes. By initially running the
+    // method outside of any destructor, we avoid any leakage problems
+    // due to #14875.
+    fn drop_all(&mut self) {
+        loop {
+            // use a separate scope to ensure that the RefCell borrow
+            // is relinquished before running `dtor`
+            let dtor = {
+                let mut dtors = self.dtors.borrow_mut();
+                if let Some(mut node) = dtors.take() {
+                    *dtors = node.next.take().map(|b| *b);
+                    node.dtor
+                } else {
+                    return
+                }
+            };
+            dtor()
+        }
+    }
+
+    pub fn defer<F>(&self, f: F) where F: FnOnce() + 'a {
+        let mut dtors = self.dtors.borrow_mut();
+        *dtors = Some(DtorChain {
+            dtor: Box::new(f),
+            next: dtors.take().map(Box::new)
+        });
+    }
+}
+
+impl<'a> Drop for Scope<'a> {
+    fn drop(&mut self) {
+        self.drop_all()
+    }
+}
+```
+
+This implementation does a few interesting things:
+
+* It avoids any allocation in the case of a single deferred computation.
```

Comment:
Ah I think my comment to this effect earlier may have been wrong as a `Box<FnBox()>` is created at least for each node in the chain.